osv/pub.py

```diff
@@ -0,0 +1,55 @@
+# Copyright 2023 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Pub version parser."""
+
+import functools
+
+from . import semver_index
+
+# Pub follows SemVer 2.0.0-rc.1:
+# https://semver.org/spec/v2.0.0-rc.1.html
+# Differences from SemVer 2.0.0 are described at
+# https://pub.dev/packages/pub_semver
+# Only difference that affects us is the inclusion of build suffix when
+# ordering.
+#  - Build suffixes are parsed the same as pre-release suffixes.
+#  - Pre-release suffixes are evaluated before build suffixes.
+#  - Versions with build suffixes come after versions without.
+#    e.g. 1.0.0-pre < 1.0.0-pre+build < 1.0.0 < 1.0.0+build
+# SemVer 2.0.0-rc.1 also does not explcitly disallow empty identifiers or
```

Comment:
Wow, nice catch! How did you notice this case?

Reply:
I was comparing 2.0.0 and 2.0.0-rc.1 side by side, and noticed the explicit callout in 2.0.0 `Identifiers MUST NOT be empty.` was not in rc.1 🙂

I noted this, but then did not actually check how our system reacts to it. The [tests in pub_semver](https://github.com/dart-lang/pub_semver/blob/master/test/version_test.dart) don't seem to test for it either.

I don't even know if Pub actually allows it, but I'll add in a test in anyway.